app/services/ajax.js

```diff
@@ -1,190 +1,162 @@
+
+    options.headers = this.setupHeaders(url, method, options);
+
+    return this.fetchRequest(url, method, options);
+  },
+
+  fetchRequest(url, method, options) {
+    return new EmberPromise((resolve, reject) => {
+      const { headers, body } = options;
+      const fetchOptions = {
+        headers,
+        method,
+      };
+      if (body) {
+        fetchOptions['body'] = body;
       }
-    }
-    if (options.contentType) {
-      xhr.setRequestHeader('Content-Type', options.contentType);
-    }
-    resolve = null;
-    reject = null;
-    promise = new EmberPromise((_resolve, _reject) => {
-      resolve = _resolve;
-      return reject = _reject;
-    });
-    xhr.onreadystatechange = () => {
-      let contentType, data;
-      if (xhr.readyState === 4) {
-        contentType = xhr.getResponseHeader('Content-Type');
-        data = (() => {
-          if (contentType && contentType.match(/application\/json/)) {
-            try {
-              return $.parseJSON(xhr.responseText);
-            } catch (error1) {
-              if (this.features.get('debugLogging')) {
-                // eslint-disable-next-line
-                console.log('error while parsing a response', method, options.url, xhr.responseText);
-              }
-            }
+
+      fetch(url, fetchOptions).then(response => {
+        if (!response.ok) {
+          this.handleFetchError(reject, response);
+        } else {
+          const { 'content-type': resContentType = '' } = response.headers.map;
+
+          let resContent;
+          if (resContentType.includes('application/json')) {
+            resContent = response.json();
           } else {
-            return xhr.responseText;
+            resContent = response.text();
           }
-        })();
-        if (xhr.status >= 200 && xhr.status < 300) {
-          resolve(data);
-          return options.success.call(options.context, data, xhr.status, xhr);
-        } else {
-          reject(xhr);
-          return options.error.call(options.context, data, xhr.status, xhr);
+
+          resContent
+            .then(data => resolve(data))
+            .catch(error => this.handleFetchError(reject, error));
         }
-      }
-    };
-    data = options.data;
-    let contentType = options.contentType;
-    let isJSON = isNone(contentType) || contentType.match(/application\/json/);
-    if (typeof options.data === 'object' && isJSON) {
-      data = JSON.stringify(data);
-    }
-    if (data) {
-      xhr.send(data);
-    } else {
-      xhr.send();
+      }).catch(error => {
+        this.handleFetchError(reject, error);
+      });
+    });
+  },
+
+  handleFetchError(reject, error) {
+    reject(error);
+    this.logFetchError(error);
+  },
+
+  logFetchError(response) {
+    if (this.features.get('debugLogging')) {
+      const { status = 'UNKNOWN' } = response;
+      // eslint-disable-next-line
+      console.log(`[ERROR] Fetch error (${status}): ${response}`);
```

Comment:
Is it worth to send it also to Sentry?

Reply:
Thanks for the review :smile: That's a good question! I've gone back and forth about it. On the one hand, yes it would be handy to have these show up in sentry and maybe help catch some issues that we would otherwise miss. But on the other, it could result in a lot of noise in sentry. I could honestly go either way, but I'd be interested to hear @artursmirnov's opinion on it

Reply:
Also for intentional logging (e.g. to make other devs aware of something) it's better to use [warn](https://api.emberjs.com/ember/3.14/functions/@ember%2Fdebug/warn) as it doesn't expose logging to production builds.

Reply:
Oooh nice I like that! and then we can also remove the eslint disable comment

Reply:
Ok, I've updated the function to use sentry and `warn` instead of `console.log`

Reply:
About the sentry, I agree, it could be too noisy. And also this general case could cover well-handled errors, like, for instance, `401` or `404` on repo pages, etc. We don't need such errors in Sentry as it's already too overwhelmed :) I'd better suggest to handle API errors on a per-case basis (where the particular endpoint is used)

Reply:
ahh yeah, good point :) I've removed the call to sentry, and left in the use of `warn`

Reply:
haha, I'm sorry for being so slow 😄

Reply:
hah no worries :D I was already looking at and working on this, so I happened to respond to it very quickly